### PR TITLE
feat(dev): reload express app on changes

### LINKED ISSUE #134: Hot reloading for the Express server
## Bug Report or Feature Request (mark with an `x`)
```
- [ ] bug report -> please search issues before submitting
- [x] feature request
```

## Desired functionality.
Hot reloading of server modules. I know the server is responsible for client and isomorphic building/HMR, so this might not be possible, but having to restart the entire stack to reflect a change in the Express code is painful.

### PR DESCRIPTION
**Pull request by Riari**
closes #134

For some reason, this approach from the article I mentioned didn't work for me:

```
if (module.hot) {
 module.hot.accept('./server', () => {
  server.removeListener('request', currentApp)
  server.on('request', app)
  currentApp = app
 })
}
```

HMR appeared to work but the changes weren't actually reflected. So I opted for this instead:

```
if (module.hot) {
  module.hot.accept();

  module.hot.dispose(data => {
    server.close();
  });
}
```

However, doing it this way seems to introduce a minor memory leak. I assume `server.close()` isn't sufficient and something else needs cleaning up too, but I'm not sure how to track that down. Will update the PR if I find a solution.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -101,6 +101,15 @@
       "pt"
     ]
   },
+  "nodemonConfig": {
```
Comment (devCrossNet) [inline package.json]: this can be removed

File: package.json [hunk 2]
```diff
@@ -184,6 +193,7 @@
     "rimraf": "2.6.2",
     "sass-loader": "7.0.1",
     "serviceworker-webpack-plugin": "1.0.0-alpha02",
+    "start-server-webpack-plugin": "^2.2.5",
```
Comment (devCrossNet) [inline package.json]: can you save the exact version?

File: src/server/dev/server.ts [hunk 3]
```diff
@@ -12,9 +12,9 @@ export default (app: Express.Application, cb: any): void => {
   let bundle: string;
   let template: string;
 
-  clientConfig.entry.app = ['webpack-hot-middleware/client', clientConfig.entry.app];
+  clientConfig.entry.server = ['webpack-hot-middleware/client', clientConfig.entry.app];
```
Comment (devCrossNet) [inline src/server/dev/server.ts]: not sure about this one. It should be `app`

**Comment by Riari** [comment]
OK, updated. For L15 in src/server/dev/server.ts, it complained when I tried renaming it:

```
[1] [HMR] WebpackOptionsValidationError: Invalid configuration object. Webpack has been initialised using a configuration object that does not match the API schema.
[1]  - configuration.entry should be one of these:
[1]    object { <key>: non-empty string | [non-empty string] } | non-empty string | [non-empty
string] | function
[1]    -> The entry point(s) of the compilation.
[1]    Details:
[1]     * configuration.entry['app'] should be a string.
[1]       -> The string is resolved to a module which is loaded upon startup.
[1]     * configuration.entry['app'][1] should be a string.
[1]       -> A non-empty string
[1]     * configuration.entry should be a string.
[1]       -> An entry point without name. The string is resolved to a module which is loaded upon startup.
[1]     * configuration.entry should be an array:
[1]       [non-empty string]
[1]     * configuration.entry should be an instance of function
[1]       -> A Function returning an entry object, an entry string, an entry array or a promise to these things.
```

It's odd that `clientConfig.entry.server` even worked. But I actually ended up completely removing that line along with the other `webpack-hot-middleware`-related lines for the client compiler, because it actually seems to work perfectly fine without it - presumably because the HMR we're adding to the Express code ultimately picks up client changes too.

Perhaps that last part is the reason behind the memory leak/excessive memory use. I did notice that whenever a change is made under `src/app`, webpack displays the following:

```
[1] i ｢wdm｣: Compiling...
[1] i ｢wdm｣: Compiling...
```

I think that might be due to restarting the server during HMR - when the server comes back up, it starts compiling everything again. Any ideas about how we might overcome that?

**Comment by devCrossNet** [comment]
I will have a look at the code tomorrow evening. Maybe a dev-server singleton could work 🤔

**Comment by devCrossNet** [comment]
Btw: thanks for spending your time on this issue 👍 I am pretty sure we‘ll find a solution 🙂

**Comment by Riari** [comment]
Maybe the best thing is to stick with the solution from the article and figure out why it's not fully working. The more I think about it, the more that makes sense vs. completely closing the server and firing it back up. Then we can put the client HMR stuff back. I'll spend some time on this tonight/tomorrow as well.

**Comment by Riari** [comment]
I've had some success by moving the SSR route stuff into the server entry point and using `server.removeListener(...)` and `server.on(...)` instead of closing the server. This seemed to fix the memory issues, but the SSR route would get lost when the app is replaced, so I tried to use some callbacks in the SSR/dev server side of things to pass along an updated renderer and re-register the route. That almost works, except any changes to the Express code cause the renderer to be out of date (you can still view frontend routes such as `/not-found`, but the URLs for the build bundles are out of date and begin to 404).

I've pushed the code to a separate branch since it deviates quite a lot from what we have here: https://github.com/Riari/vue-starter/commit/aec3584e525c748d1b13fd4f42165c2745e1a0bf

I'm not really sure where to go from there. We need an elegant way of retaining the SSR route when the Express code changes, whilst also making sure it always uses the current renderer. 🤔

**Comment by devCrossNet** [comment]
@Riari can you please have a look at this commit: https://github.com/devCrossNet/vue-starter/commit/8fb5394988b9467987a5560a3a8acfb7eafe585b

I think I got a stable solution. the key was to make the dev-server some kind of a singleton. I don't know how you profiled the memory leak, can you please do the same profiling with this solution?

I tested it now ca. 30 minutes and it worked stable and the changes reflected very quickly.

Thanks in advance.

**Comment by devCrossNet** [comment]
@Riari If this is a solution you like, please update your PR. I would like to merge it as part of the v1 milestone 😊

**Comment by Riari** [comment]
Nice, that works perfectly 👍 memory usage appears to be fine too (it remains below 300 MB for me).

This project has been a huge help btw, I appreciate all of the work you've put into it!

**Comment by Riari** [comment]
I ended up having to regenerate package-lock.json after rebasing to solve some dependency issues, but I think that's fine, especially as immediate dependency versions are locked down in package.json.

**Comment by devCrossNet** [comment]
@Riari regenerating the package-lock is totally fine, I do this on a regular basis 👍 

I don't get a lot of feedback, thank you very much!

hopefully, the contributing process was ok and I see some more in the future :-)

**Pull request merged**